react/features/whiteboard/middleware.ts

```diff
@@ -0,0 +1,201 @@
+ * @param {Object} attributes - The command attributes.
+ * @returns {void}
+ */
+const handleWhiteboardStatus = (store: IStore, { roomId, roomKey }: { roomId: string; roomKey: string; }) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const username = getLocalParticipant(state)?.name;
+    const isOpen = isWhiteboardOpen(state);
+
+    if (isOpen) {
+        return;
+    }
+
+    localStorage.setItem(WHITEBOARD_STORAGE_KEY, JSON.stringify({ username }));
+    focusWhiteboard(store);
+    dispatch(setupWhiteboard({
+        collabDetails: {
+            roomId,
+            roomKey
+        }
+    }));
+};
+
+/**
+ * Sends the WHITEBOARD command.
+ *
+ * @returns {void}
+ */
+const sendWhiteboardCommand = ({ conference, roomId, roomKey }: {
+    conference: any;
+    roomId?: string;
+    roomKey?: string;
+}): void => {
+    conference.sendCommand(WHITEBOARD, {
+        attributes: {
+            roomId,
+            roomKey
+        }
+    });
+};
+
+/**
+ * Middleware which intercepts whiteboard actions to handle changes to the related state.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register((store: IStore) => (next: Function) => async (action: any) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const localParticipant = getLocalParticipant(state);
+
+    switch (action.type) {
+    case OPEN_WHITEBOARD: {
+        if (!isParticipantModerator(localParticipant) && !isEveryoneModerator(state)) {
+            return;
+        }
+
+        const conference = getCurrentConference(state);
+        const collabDetails = await generateCollaborationLinkData();
+
+        dispatch(setTileView(false));
+        sendWhiteboardCommand({
+            ...collabDetails ? {
+                roomId: getCurrentRoomId(state),
+                roomKey: collabDetails?.roomKey
+            } : {},
+            conference
+        });
+
+        break;
+    }
+    case CONFERENCE_JOIN_IN_PROGRESS: {
+        const { conference } = action;
+
+        // @ts-ignore
+        conference.addCommandListener(WHITEBOARD, ({ attributes }) => {
+            handleWhiteboardStatus(store, attributes);
+        });
+        break;
+    }
+
+    // TODO: cleanup PARTICIPANT_DISPLAY_NAME_CHANGED
+    case PARTICIPANT_UPDATED: {
+        if (typeof APP !== 'undefined') {
+            const participant = getLocalParticipant(store.getState());
+
+            if (participant
+                && participant.id === action.participant.id
+                && action.participant.name
+                && participant.name !== action.participant.name) {
+                localStorage.setItem(WHITEBOARD_STORAGE_KEY, JSON.stringify({ username: action.participant.name }));
+            }
+        }
+        break;
+    }
+    case PIN_PARTICIPANT: {
+        if (action.participant.id !== WHITEBOARD_ID) {
+            blurWhiteboard(store);
```

Comment:
Why do we need to explicitly remove the whiteboard from the stage filmstrip if the user pins another participant? Doesn't the pinning system take care of that automagically?

Reply:
I received feedback that the whiteboard keeps taking the stage, but I can't repro anymore, so I'll drop this part.